KAFKA-13668: Retry upon missing initProducerId due to authorization error

philipnee opened the pull request:
*Currently producer will throw a fatal error upon failing `initProducerId`, which can be caused by authorization errors. In this case, the user will need to instantiate a producer.

The goal for this PR is to make authorization errors non-fatal so that the user can retry until the permission is fixed by an admin.

Here we first transition the producer to the `ABORTABLE` state, then to the `UNINITIALIZED` state (so that the producer is recoverable). Upon the subsequent send, the producer will transition to `INITIALIZING` and attempt to send another InitProducerIdRequest.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -1292,8 +1292,10 @@ public void handleResponse(AbstractResponse response) {
-                    error == Errors.CLUSTER_AUTHORIZATION_FAILED) {
+            } else if (error == Errors.CLUSTER_AUTHORIZATION_FAILED) {
+                log.warn("Retrying upon missing initProducerId due to cluster authorization error");
+                reenqueue();
```
Hmm, I think we do want the error to get propagated back to the application. Most users would expect the application to fail so that they have a chance to see the issue and fix it. High level, what I was thinking we could do is fail the request by calling `result.fail`, then transition back to `UNINITIALIZED`. Then if the user wants to, they can call `initTransactions()` to retry.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -1292,8 +1292,13 @@ public void handleResponse(AbstractResponse response) {
+                log.error("Cluster authorization failure, transition the producer state back to uninitialized");
+                fail(error.exception());
+                // transition back to uninitialized state to enable client side retry
+                transitionTo(State.READY);
```
Could we update `isTransitionValid` to allow transitions from INITIALIZING to UNINITIALIZED?

ijuma commented:
@philipnee where are we with this?

philipnee commented:
@ijuma - I think @hachikuji is reviewing it.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -1292,8 +1296,12 @@ public void handleResponse(AbstractResponse response) {
+                fail(error.exception());
+                // transition back to uninitialized state to enable client side retry
+                transitionTo(State.UNINITIALIZED);
+            } else if (error == Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED) {
```
I wonder if we should handle this case consistently. I cannot think of any reason not to.

philipnee replied:
yeah that sounds right. it seems like you they should just configure the ACL.

hachikuji commented on clients/src/test/resources/log4j.properties:
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
-log4j.rootLogger=OFF, stdout
+log4j.rootLogger=ALL, stdout
```
Probably added by mistake?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -627,6 +627,9 @@ private TransactionManager configureTransactionState(ProducerConfig config,
      * @throws KafkaException if the producer has encountered a previous fatal error or for any other unexpected error
      * @throws TimeoutException if the time taken for initialize the transaction has surpassed <code>max.block.ms</code>.
      * @throws InterruptException if the thread is interrupted while blocked
+     * @throws org.apache.kafka.common.errors.ClusterAuthorizationException fatal error indicating that the idempotent
```
I think we can get rid of this since we have `AuthorizationException` above. Perhaps we can add the comment about retrying there?

hachikuji reviewed (approved):
Thanks, LGTM.

philipnee committed:
KAFKA-13668: sender abort on cluster or trasactionID auth error

Tests and impl

philipnee committed:
KAFKA-13668: sender abort on cluster or trasactionID auth error

Tests and impl

philipnee committed:
This is not needed.

philipnee committed:
This is not needed.

philipnee committed:
unused imports

philipnee committed:
unused imports

philipnee committed:
should have thrown abortable error

philipnee committed:
should have thrown abortable error

philipnee commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -607,6 +608,14 @@ public synchronized void handleCompletedBatch(ProducerBatch batch, ProduceRespon
+    public synchronized void transitionToUninitialized(RuntimeException exception) {
+        transitionTo(State.UNINITIALIZED);
+        lastError = null;
+        if (pendingTransition != null) {
```
I'm actually not sure if this is needed.

jolshan replied:
Do you have more background on what pendingTransitions we typically have? I don't think this hurts unless it will cause a fatal error. If it does, we may need to take a closer look.

philipnee replied:
thanks, i think it's fine to leave it here.

jolshan replied:
I'd prefer you confirm that it's fine. Looking at the code -- it seems to set the TransactionalRequestResult state to the error and count down the latch. Given that it is for the request -- looks to be fine.

philipnee committed:
clean up on naming

philipnee committed:
clean up on naming

philipnee committed:
Fix a test

philipnee committed:
Fix a test

philipnee commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -1237,6 +1238,35 @@ public void testInitTransactionWhileThrottled() {
+            assertThrows(ClusterAuthorizationException.class, () -> producer.initTransactions());
+            // retry initTransactions after the ClusterAuthorizationException not being thrown
+            client.prepareResponse(initProducerIdResponse(1L, (short) 5, Errors.NONE));
+            producer.initTransactions();
```
hmm... consistently having issue to get a Errors.NONE response.  I can sometimes get a ClusterAuthorizationException.  Not sure why is it

laosiaudi commented:
Following this PR because of https://github.com/confluentinc/kafka-rest/issues/416 and https://issues.apache.org/jira/browse/KAFKA-13668. After this is fixed, does it mean idempotent producer is supported on Kafka-rest?  One more follow-up question is, does idempotent producer on Kafka-rest handle the connection failure between user and Kafka-rest well? I guess it's not?

philipnee committed:
Merge branch 'KAFKA-13668-retry-upon-missing-initProducerId-for-idmpt-producer' of github.com:philipnee/kafka into KAFKA-13668-retry-upon-missing-initProducerId-for-idmpt-producer

philipnee committed:
Merge branch 'apache:trunk' into KAFKA-13668-retry-upon-missing-initProducerId-for-idmpt-producer

philipnee committed:
clean up

philipnee committed:
lint

philipnee committed:
check abortable error

philipnee committed:
re-test

philipnee committed:
retry until succeed

philipnee committed:
abortable exception

philipnee commented:
Hey @hachikuji - Thanks for the previous review, but I rewrote the PR because of some of the timeout IT tests.  In the previous version, the issue was that simply setting the transactionManager state to UNINITIALIZED wouldn't propagate the authorization exceptions to the sender, so the producer would keep re-requesting the producerID, eventually causing some tests to timeout. Here in this PR, I transition to the "ABORTABLE_ERROR" state, abort the pending requests, then send the state to UNINITIALIZED in the sender.

Do you agree with this approach?

Also, do you think we need to make this behavior consistent with the produce request? For example, the tests in `SenderTests.java`:
- `testClusterAuthorizationExceptionInProduceRequest`
- `testClusterAuthorizationExceptionInInitProducerIdRequest`

kirktrue reviewed (commented):
Can you update the pull request summary to include a... summary?

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -300,9 +301,13 @@ void runOnce() {
 
+                RuntimeException lastError = transactionManager.lastError();
+                if (transactionManager.hasAbortableError() && shouldHandleAuthorizationError(lastError)) {
+                    return;
```
In the new code we return in the case of a `ClusterAuthorizationException`. Per the comments later on in the `runOnce` method, do we still want to call the `bumpIdempotentEpochAndResetIdIfNeeded()` method in this case?

philipnee replied:
I'm thinking not, because we aren't adding a new producer.  @jolshan thoughts?

jolshan replied:
Sorry -- I think if we have an error we'd still want to bump the epoch right?

jolshan replied:
(Assuming this is an indempotent producer and not a transactional one)

philipnee replied:
Right this is an idempotent producer, thanks for catching the epoch bump, I was wondering about that too.

jolshan replied:
Did we determine that we will end up bumping the epoch? It doesn't look like it is done here so is it somewhere else?

philipnee replied:
Hey @jolshan - I actually think we don't need to manually bump the epoch here.  I think it is already handled by the existing logic.  Here's the explanation. I think initProducerId is only happening at 2 different places: 1. `initializeTransactions` and 2. in the sender loop for idempotent producer `bumpIdempotentEpochAndResetIdIfNeeded`.

for 1. It will bump the epoch if the epoch != None, which means the producer has been initialized and needs to bump the epoch upon re-requesting the id per your comment

for 2. it is when we first initialize a producer (so it doesn't have an id at the beginning), and the `InitProducerIdRequest` should bump the epoch to 0 upon first successful attempt.  This is the case we are addressing in this PR I think.

philipnee replied:
FWIW: I think in the context of this ticket, we are trying to avoid poisoning the client when the client is unable to re-authenticate upon startup. The fix wants to continue to retry the request until the permission is fixed. So the handler should handle the epoch bump.

jolshan replied:
Thanks Philip -- I think I forgot that this was the initProducerId call -- so we don't really have an epoch yet. 😅 We set to 0 after getting the producer ID.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -996,20 +1010,21 @@ private void ensureTransactional() {
     }
 
     private void maybeFailWithError() {
```
Are there any functional changes in `maybeFailWithError`, or is it just to refactor for an early return?

philipnee replied:
no change, i just refactored it a bit.

kirktrue reviewed (commented):
Thanks for the PR, @philipnee. Unfortunately I know next-to-nothing about the transaction manager, so my review is non-binding.

dajac commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -2236,6 +2236,30 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
     assertFalse(response.clusterId.isEmpty, "Cluster id not returned")
   }
 
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
```
nit: Indentation is off for this entire test.

philipnee committed:
indentation fix

jolshan commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -618,7 +618,8 @@ private TransactionManager configureTransactionState(ProducerConfig config,
      *         does not support transactions (i.e. if its version is lower than 0.11.0.0)
      * @throws org.apache.kafka.common.errors.AuthorizationException fatal error indicating that the configured
-     *         transactional.id is not authorized. See the exception for more details
+     *         transactional.id is not authorized, or the idempotent producer id is unavailable. See the exception for
```
What do we mean by idempotent producer ID is unavailable? Is it that the broker hosting the transaction coordinator is unavailable? And do we return this error if the broker is not available?

jolshan replied:
Should we also change this to not say "fatal error"

philipnee replied:
One could get Authorization issue upon requesting for a producerId, not necessarily caused by an unavailable broker. Yes, maybe remove the word "fatal" thanks!

jolshan replied:
Can we remove fatal here still?

philipnee replied:
🤦 - sorry and done.

jolshan commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -155,7 +155,7 @@ private enum State {
             switch (target) {
                 case UNINITIALIZED:
-                    return source == READY;
+                    return source == READY || source == ABORTABLE_ERROR;
```
Do we know if there are other implications for this state machine change? It seems sort of like we want to just make the authorization error retriable -- but instead we are manking it abortable + making this specific abortable error reset to "uninitialized"

I'm not an expert in this area, but are there any other implications with re-initializing the producer?

philipnee replied:
It's one way (that I could think of) to propagate the authorization error from the sender loop; otherwise, the sender will continue to retry and causes timeout in some cases.

jolshan replied:
Hmmm. I see. I'm just wondering if we are hijacking the current state machine in an unexpected way and if there are implications there. I suppose we are only following this path on these specific error types, but I wonder if we are missing anything existing by changing the valid transitions and/or opening up the potential for something else in the future.

philipnee replied:
hmm good point, I guess upon re-initializing (transition from UNINITIALIZED to INITIALIZING state), should we check the previous error to ensure a valid transition? Maybe in `initializeTransactions` we examine the previous error and make the next transition?

jolshan replied:
Maybe I'm missing something here, but from "abortable_error" don't we typically abort the transaction and then return to initializing? Sorry if you've explained this before, but why did we choose uninitialized here?

philipnee replied:
I think you will need to reinstantiate the producer if it is in an aborting state. I think here i'm trying to bring it back to initializing state so that it can issue another producerId request.

jolshan replied:
Can you point to me where you saw this? I think abortable error was something we had before that didn't need this extra step.

philipnee replied:
We need to manually transition the state back to initializing no? I think otherwise it get stucks in an error state here

https://github.com/apache/kafka/blob/a2c9f421af40e0c8ace722be94aedf8dec4f2b31/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L998

jolshan replied:
Ack -- I wonder if there are any other location where we would not want this transition. I can't think of off the top of my head, but I do think we need to be careful to clean the state properly in these cases.

jolshan commented on clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java:
```diff
@@ -1237,6 +1238,34 @@ public void testInitTransactionWhileThrottled() {
+
+        // retry initTransactions after the ClusterAuthorizationException not being thrown
+        client.prepareResponse(initProducerIdResponse(1L, (short) 5, Errors.NONE));
+        TestUtils.retryOnExceptionWithTimeout(1000, 100, producer::initTransactions);
```
To be clear -- since I couldn't tell fully from the description -- before this change, trying to call initTransactions here would fail since the producer hits a fatal state. But with this change -- we don't retry initProducerId but instead keep the producer alive and basically re-initialize it so subsequent initTransactions calls can be made (and succeed).

philipnee replied:
yes!

jolshan replied:
Are we fully validating the request is successful? Is that what the TestUtils method does here?

philipnee replied:
right, it retries initTransaction until successful or timeout.

jolshan replied:
Should we close the producer here? Just looking through the failed tests and wanted to close any gaps we may have.

philipnee replied:
Probably - it's just done very consistently in the tests 😅

jolshan commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -2236,6 +2236,30 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+    removeAndVerifyAcls(Set(denyWriteAce), wildcard)
+    addAndVerifyAcls(Set(allowWriteAce), prefixed)
+    addAndVerifyAcls(Set(allowWriteAce), literal)
+    val future = producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, "hi".getBytes))
```
Ah this test makes the benefit a bit more clear to me -- a subsequent send call works just fine.

philipnee committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-13668-retry-upon-missing-initProducerId-for-idmpt-producer

philipnee committed:
set epoch bump require

philipnee commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -1294,7 +1309,10 @@ public void handleResponse(AbstractResponse response) {
-                fatalError(error.exception());
+                log.info("Abortable authorization error: {}.  Transition the producer state to {}", error.message(), State.ABORTABLE_ERROR);
+                lastError = error.exception();
+                epochBumpRequired = true;
```
bumping epoch here - I think sender should bump it on the next iteration.

jolshan replied:
can we confirm this works as intended via test?

philipnee replied:
hey sorry - i'm in progress of writing a test here 😅

jolshan replied:
Was about to ask about this and realized we removed the bump. I guess we could still have a test unless we think the other testing covers this end to end.

philipnee replied:
even i think the initProducerId path in the sender has been tested, but it's probably worth adding a sender test to make sure the epoch is bumped from -1 to 0. I think it should be an easy modification, 👀

philipnee replied:
@jolshan  - See the last commit.  I cleaned up the comments because authorization isn't a fatal error after the patch.  I also patched a test to demonstrate resend after fixing the permission should bring the producer back to normal. Also tested the epoch there to ensure it is 0, not -1 after fixing the permission.

philipnee committed:
we actually don't need to bump the epoch..

philipnee committed:
not fatal

philipnee committed:
clean up

philipnee committed:
Update SenderTest.java

jolshan commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -2175,7 +2182,7 @@ public void testClusterAuthorizationExceptionInProduceRequest() throws Exception
 
-        // cluster authorization errors are fatal, so we should continue seeing it on future sends
+        // expecting to continue to see authorization error until user permission is fixed
         assertTrue(transactionManager.hasFatalError());
```
Should this still be true? Do we remove it from fatal error state?

philipnee replied:
I think that's actually my mistake.  It's a produce error not initProducerId.  I'll correct these issues. Sorry.

jolshan replied:
Ah I also missed. Thanks for correcting

jolshan commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -2189,7 +2196,7 @@ public void testCancelInFlightRequestAfterFatalError() throws Exception {
         assertTrue(transactionManager.hasProducerId());
 
-        // cluster authorization is a fatal error for the producer
+        // expecting authorization error on send
```
ditto here -- should we change this test to use a fatal error?

philipnee committed:
reverted some of the comment change from fatal to non-fatal

revert some of the documentation changes as Cluster Authorization can be fatal on send.

philipnee committed:
Update SenderTest.java

philipnee commented:
Hey @jolshan - Thanks for the review.  i reverted those documentation/comment changes in the senderTest.java (for the produceResponse authorization error).

philipnee commented:
Failing tests don't seem to be related.

```
Build / JDK 8 and Scala 2.12 / testReplicationIsCreatingTopicsUsingProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
1m 46s
Build / JDK 8 and Scala 2.12 / testSyncTopicConfigUseProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 26s
Build / JDK 8 and Scala 2.12 / testSyncTopicConfigUseProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
1m 46s
Build / JDK 11 and Scala 2.13 / testReplicationIsCreatingTopicsUsingProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
3m 15s
Build / JDK 11 and Scala 2.13 / testCreatePartitionsUseProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 3s
Build / JDK 11 and Scala 2.13 / testSyncTopicConfigUseProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 41s
Build / JDK 11 and Scala 2.13 / testReplicationIsCreatingTopicsUsingProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 23s
Build / JDK 11 and Scala 2.13 / testCreatePartitionsUseProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
1m 55s
Build / JDK 11 and Scala 2.13 / testSyncTopicConfigUseProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 7s
Build / JDK 11 and Scala 2.13 / testWithGroupMetadata() – kafka.api.TransactionsBounceTest
1m 22s
Build / JDK 11 and Scala 2.13 / testCreateClusterAndCreateAndManyTopics() – kafka.server.KRaftClusterTest
18s
Build / JDK 17 and Scala 2.13 / testReplicationIsCreatingTopicsUsingProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 14s
Build / JDK 17 and Scala 2.13 / testCreatePartitionsUseProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 22s
Build / JDK 17 and Scala 2.13 / testSyncTopicConfigUseProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 16s
Build / JDK 17 and Scala 2.13 / testReplicationIsCreatingTopicsUsingProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 16s
Build / JDK 17 and Scala 2.13 / testCreatePartitionsUseProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 53s
Build / JDK 17 and Scala 2.13 / testSyncTopicConfigUseProvidedForwardingAdmin() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
2m 13s
Build / JDK 17 and Scala 2.13 / testGetSinkConnectorOffsetsDifferentKafkaClusterTargeted – org.apache.kafka.connect.integration.OffsetsApiIntegrationTest
1m 14s
Existing failures - 3
Build / JDK 8 and Scala 2.12 / putTopicStateRetriableFailure – org.apache.kafka.connect.storage.KafkaStatusBackingStoreFormatTest
<1s
Build / JDK 11 and Scala 2.13 / putTopicStateRetriableFailure – org.apache.kafka.connect.storage.KafkaStatusBackingStoreFormatTest
<1s
Build / JDK 17 and Scala 2.13 / putTopicStateRetriableFailure – org.apache.kafka.connect.storage.KafkaStatusBackingStoreFormatTest
```

jolshan commented:
Going to rerun the build one more time.

jolshan commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -328,6 +333,21 @@ void runOnce() {
+    // We handle {@code TransactionalIdAuthorizationException} and {@code ClusterAuthorizationException} by first
+    // failing the inflight requests, then transition the state to UNINITIALIZED so that the user doesn't need to
+    // instantiate the producer again.
+    private boolean shouldHandleAuthorizationError(RuntimeException exception) {
```
Just curious -- if we get an auth error on another request (ie, not initProducerId) do we expect to start over by initializing with a new ID? 

Also what is the goal with the poll call? Is it just replacing line 308? Would the code work without it?

philipnee replied:
it seems like all of the non-initProducerId TransactionalIdAuthorizationException and ClusterAuthorizationException are fatal.

For the poll: I think we don't need it because there's no outbound request, as it should've been already polled in the previous `runOnce`.  The tests seem to work without so i'll remove it.

philipnee replied:
I might have added it there because a few lines down, in the fatal error handling, the client was polled before return.

jolshan commented:
Left a few more questions -- I think we are in the final stretch here.

philipnee committed:
we don't need the poll

jolshan commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -300,9 +301,13 @@ void runOnce() {
+                if (transactionManager.hasAbortableError() && shouldHandleAuthorizationError(lastError)) {
+                    return;
+                }
+
```
I thought I left this comment, but seems like it didn't take -- should we move the fatal error check above the auth check? Since for other request types, we will go through the above path, but it really should just be a fatal error.

philipnee replied:
sounds good - perhaps you meant by checking for fatal first then abortable error? Though I think it doesn't change the logic there.

jolshan replied:
Yup -- my concern is we unecessarily reset the producer to initializing in the fatal error case.

jolshan replied:
Ah I see we had the abortable error check. 😅 Ok well now we are doubly covered.

philipnee committed:
check fatal first

jolshan reviewed (approved):
Running the tests one more time, but otherwise, should be good to go.

jolshan commented:
Still an issue -- rebuilding again 😅

philipnee commented:
thank you~

philipnee commented:
There seem to be a bit of server related failures, but they are also irrelevant to this change I think:

I believe JDK17 tests and most of the JDK11 tests passed.

```
Build / JDK 11 and Scala 2.13 / testOffsetSyncsTopicsOnTarget() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
3m 24s
Build / JDK 8 and Scala 2.12 / testSyncTopicConfigs() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationTransactionsTest
1m 51s
Build / JDK 8 and Scala 2.12 / testLogCleanerConfig(String).quorum=kraft – kafka.server.DynamicBrokerReconfigurationTest
28s
Build / JDK 8 and Scala 2.12 / [1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.zk.ZkMigrationIntegrationTest
2m 7s
Existing failures - 23
Build / JDK 8 and Scala 2.12 / executionError – kafka.server.DynamicBrokerReconfigurationTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.EdgeCaseRequestTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.FetchRequestDownConversionConfigTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.FetchRequestTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.FetchRequestWithLegacyMessageFormatTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.FinalizedFeatureChangeListenerTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.GssapiAuthenticationTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.KafkaMetricReporterClusterIdTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.KafkaMetricReporterExceptionHandlingTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.KafkaMetricsReporterTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.KafkaServerTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.ListOffsetsRequestTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.LogDirFailureTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.LogRecoveryTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.MetadataRequestTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.MultipleListenersWithAdditionalJaasContextTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.OffsetFetchRequestTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.ServerGenerateClusterIdTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.ServerShutdownTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.UpdateFeaturesTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.server.epoch.EpochDrivenReplicationProtocolAcceptanceWithIbp26Test
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.zk.KafkaZkClientTest
<1s
Build / JDK 8 and Scala 2.12 / initializationError – kafka.zk.ZkMigrationClientTest
<1s
```

philipnee committed:
closing the producer

jolshan commented:
Thanks Philip. The initialization ones come when there is a thread leak somewhere. I'm pretty convinced it wasn't your change, but just wanted to be extra safe. I will check the next build. Thanks for your patience.
